packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,335 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { TypedEventEmitter, assert } from "@fluidframework/common-utils";
+import { IConnectionDetails, IDeltaHandlerStrategy, IDeltaQueue, IDeltaQueueEvents } from "@fluidframework/container-definitions";
+import { IDocumentService } from "@fluidframework/driver-definitions";
+import { IDocumentMessage, ISequencedDocumentMessage, MessageType } from "@fluidframework/protocol-definitions";
+import { IConnectionManager, IConnectionManagerFactoryArgs } from "./contracts";
+import { DeltaManager } from "./deltaManager";
+
+const GroupedBatchOpType = "groupedBatch";
+
+/**
+ * Underneath, stores exactly the messages that are going to/from service.
+ * Key point is that anyone who wants to observe the events or see the content of the DeltaQueue
+ * will see the version that is virtualized (not what is sent over the wire).
+ *
+ * For example:
+ * - Say we have ops 1, 2, 3 and they get grouped together in a new "groupedBatch" op
+ * - Over the wire, we send just one op and the DeltaQueue will store just this single op
+ * - When other places subscribe to "push"/"op" event or call "toArray"/"peek", they will see a version that
+ * is inline with when the ops 1, 2, 3 were separate ops and will not see the "groupedBatch" op
+ */
+class VirtualDeltaQueueProxy<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> implements IDeltaQueue<T> {
```

Comment:
Overall observation RE exposure of queues & leakage:
1. Looking at usage of inboud queue in DM, I think we can remove it from there by adding couple abstract members, and not implementing IDQ interface (i.e. pushing implementation of public interface to be only on virtualized class). That way, inbound is projected only down (to runtime), and it's semantics are very clear. 
2. outboud is tough. It's implemented by ConnectionManager. DM never touches it in any way. But I think we can get rid of this concept altogether in ConnectionManager. Basically, queue should be only used when we have a connection. And in such case, it's not really a queue, as messages are pushed right away to socket. It's async today (queue processes items async, so it does work a bit like a queue, i.e. it could accumulate items), but it does not have to - there are no issues with sending ops right away (same can't be said for inboud queue!). I wonder, if we could prototype getting rid of outbound queue concept in ConnectionManager as a separate PR?
   - I see the only place where we use it in runtime - see connectionTelemetry.ts. And in there, we assume we see non-processed ops right before we send them. This telemetry measures how long it takes from op generation (on runtime side) till it actually gets sent. But with batching accumulation on runtime now, we should be able to rely instead on ContainerRuntime.flush() providing these signals

Reply:
One addition - it's a a lot of processing (packing/unpacking) for "push" flow on oubound queue. It's very inefficient (from memory consumption / CPU consumption POV), so avoiding it (by ensuring runtime does not need to rely on it) would be great.
It should be possible to change that in separate PR - I'd not try to jam everything into here (and it's good change to make regardless). Please let me know if you need help - I'm more familiar with that code and can help.